### ISSUE TITLE
Reproduction discrepancy between the paper and the released implementation

### ISSUE THREAD

**Issue by Zxl19990529** [hanxunyu/VisionTrim#10]
# Reproduction discrepancy between the paper and the released implementation

Hi, thank you for releasing VisionTrim.

We are trying to reproduce Table 1 on LLaVA-1.5-7B using the official evaluation protocols. During the reproduction, we found several differences between the method described in the paper and the current public implementation. Could you please clarify which implementation and configuration were used to produce Table 1?

## 1. LTAM and adaptive weighting appear to be overwritten

In Section 3.1, Eq. (6), the paper combines global importance and LTAM local importance using adaptive variance-based weighting:

$$
S_i = \alpha \hat{S}_i^g + (1-\alpha) S_i^l
$$

In [`llava/model/multimodal_encoder/clip_encoder.py`](https://github.com/hanxunyu/VisionTrim/blob/308e73a9b8437882c5c1056a32dbc60fd41efd53/llava/model/multimodal_encoder/clip_encoder.py#L438-L444), the code first computes:

```python
metric = combine_importance(
    cls_attn,
    local_token_importance,
    method="adaptive",
    alpha=0.5,
)
```

However, this result is immediately overwritten by:

```python
alpha = 0.5
metric = alpha * cls_attn_norm + (1 - alpha) * avg_received_attn_norm
```

Therefore, the metric passed to `torch.topk` does not use the value returned by `combine_importance`. Neither `local_token_importance` nor its variance-derived weights affect this `topk` call; the selected metric instead uses a fixed weight and `avg_received_attn`.

Is the second assignment unintended, or was this alternative metric used for the results in Table 1?

## 2. Please clarify the TGVC implementation relative to Eqs. (7)–(10)

The paper describes:

- softmax-normalized text–vision similarity;
- assignment scores based on text-guided similarities;
- similarity-weighted cluster aggregation.

We could not directly map the released implementation to these equations. The code appears to use text-guided CLIP similarities followed by CLIP `logit_scale`, $L_2$ normalization, hard `argmax` assignment, and count-based averaging:

```python
metric_normalized = metric_1 / metric_1.norm(dim=-1, keepdim=True)
assign_one_hot.scatter_(
    2,
    metric_normalized.argmax(dim=2).unsqueeze(-1),
    1,
)
counts = assign_one_hot.sum(dim=1).clamp(min=1).unsqueeze(-1)
aggregated_hidden = (
    torch.matmul(assign_one_hot.transpose(1, 2), hidden_to_merge)
    / counts
)
```

Could you confirm whether this is the exact TGVC variant used in Table 1?

## 3. The within-LLM pruning stage does not appear to execute

Section 3.3 describes a two-stage strategy:

1. pruning in the vision encoder;
2. dynamic DVTS/TGVC pruning inside the LLM.

Table 26 further specifies:

- pre-LLM retention rate: $50\%$;
- within-LLM insertion layer: the 2nd layer;
- dynamically adjusted within-LLM retention rate.

However, in [`llava/model/language_model/llava_llama.py`](https://github.com/hanxunyu/VisionTrim/blob/308e73a9b8437882c5c1056a32dbc60fd41efd53/llava/model/language_model/llava_llama.py#L525-L686), the decoder-side token-reranking and merging block appears to be fully commented out. The active forward path directly executes each decoder layer without shortening the visual-token sequence.

As a result, in the released execution path, passing `DVTS_token_num=48` and `TGVC_token_num=16` produces 64 visual tokens directly from the vision tower, with no second-stage pruning inside the LLM.

Could you clarify whether Table 1 used:

- only the released pre-LLM implementation; or
- the complete two-stage implementation described in Section 3.3?

If the latter was used, could you release the corresponding implementation?

## 4. Ambiguity between Table 11 and Table 26

Table 11 defines the 64-token setting as:

- DVTS: 48 tokens;
- TGVC: 16 tokens.

This matches the released evaluation scripts.

Table 26, however, states that the pre-LLM stage retains $50\%$ of the tokens and that another dynamically adjusted pruning stage is inserted at the second LLM layer.

It is unclear how these two configurations are combined to obtain the “Retain Averaged 64 Tokens” result in Table 1.

Could you provide the exact token counts before and after each pruning stage for the 64/128/192-token settings?

## 5. Reproduction results using the released execution path

Using the released model entrypoints, greedy decoding, the documented $3\!:\!1$ DVTS/TGVC split, and the native local scorers, we obtained:

| Tokens | GQA | MME | POPE | ScienceQA | TextVQA |
|---:|---:|---:|---:|---:|---:|
| 64 | 53.24 | 1627.44 | 70.20 | 69.84 | 55.09 |
| 128 | 56.38 | 1736.46 | 79.81 | 69.82 | 56.52 |
| 192 | 58.24 | 1765.19 | 83.69 | 69.77 | 57.37 |

The corresponding Table 1 values are:

| Tokens | GQA | MME | POPE | ScienceQA | TextVQA |
|---:|---:|---:|---:|---:|---:|
| 64 | 58.8 | 1780 | 86.2 | 71.0 | 56.8 |
| 128 | 60.3 | 1788 | 86.6 | 69.7 | 58.2 |
| 192 | 61.0 | 1798 | 86.8 | 70.8 | 58.4 |

In our local reproduction setup, we verified that the evaluated samples are complete and that multi-GPU chunking, concatenation order, prompts, token-budget arguments, greedy decoding, and local scorer inputs are consistent.

## Requested reproduction information

Would it be possible to provide the following?

1. The exact commit used for Table 1.
2. The decoder-side pruning implementation, if it was used for Table 1.
3. The exact LTAM/DVTS metric used for Table 1.
4. The exact TGVC assignment and aggregation implementation.
5. The token count after the pre-LLM and within-LLM stages.
6. The Python, PyTorch, Transformers, and Accelerate versions.
7. The exact LLaVA-1.5-7B checkpoint revision.
8. The evaluation data and scorer version corresponding to Table 1.

Thanks for your help.
